src/library.js

```diff
@@ -2223,43 +2223,61 @@ mergeInto(LibraryManager.library, {
 
   // random.h
 
-  // TODO: consider allowing the API to get a parameter for the number of
-  // bytes.
-  $getRandomDevice: function() {
+  $initRandomDevice: function() {
     if (typeof crypto == 'object' && typeof crypto['getRandomValues'] == 'function') {
       // for modern web browsers
-      var randomBuffer = new Uint8Array(1);
-      return () => { crypto.getRandomValues(randomBuffer); return randomBuffer[0]; };
+#if SHARED_MEMORY
+      // like with most Web APIs, we can't use Web Crypto API directly on shared memory,
+      // so we need to create an intermediate buffer and copy it to the destination
+      return (view) => (
+        view.set(crypto.getRandomValues(new Uint8Array(view.byteLength))),
+        // Return the original view to match modern native implementations.
+        view
+      );
+#else
+      return (view) => crypto.getRandomValues(view);
+#endif
     } else
 #if ENVIRONMENT_MAY_BE_NODE
     if (ENVIRONMENT_IS_NODE) {
       // for nodejs with or without crypto support included
       try {
         var crypto_module = require('crypto');
-        // nodejs has crypto support
-        return () => crypto_module['randomBytes'](1)[0];
+        var randomFillSync = crypto_module['randomFillSync'];
+        if (randomFillSync) {
+          // nodejs with LTS crypto support
+          return (view) => crypto_module['randomFillSync'](view);
+        }
+        // very old nodejs with the original crypto API
+        var randomBytes = crypto_module['randomBytes'];
+        return (view) => (
+          view.set(randomBytes(view.byteLength)),
+          // Return the original view to match modern native implementations.
+          view
+        );
       } catch (e) {
         // nodejs doesn't have crypto support
       }
     }
 #endif // ENVIRONMENT_MAY_BE_NODE
     // we couldn't find a proper implementation, as Math.random() is not suitable for /dev/random, see emscripten-core/emscripten/pull/7096
 #if ASSERTIONS
-    return () => abort("no cryptographic support found for randomDevice. consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };");
+    abort("no cryptographic support found for randomDevice. consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };");
 #else
-    return () => abort("randomDevice");
+    abort("initRandomDevice");
 #endif
   },
 
-  getentropy__deps: ['$getRandomDevice'],
+  $randomFill__deps: ['$initRandomDevice'],
+  $randomFill: function(view) {
+    // Lazily init on the first invocation.
+    return (randomFill = initRandomDevice())(view);
```

Comment:
Did you verify that this lazy init approach is doing the right thing?

Reply:
Hm, just verified just in case, but yeah, it's called only once & returns valid value. I mean, why wouldn't it? It's just replacing the function binding with another function. I think I saw this even recommended in js-library docs at some point.

Reply:
> I think I saw this even recommended in js-library docs at some point.

Ah that was slightly different use-case (instant initialization instead of first access), but principle is the same:

![image](https://user-images.githubusercontent.com/557590/227375077-992f30a4-f084-46d0-977d-b1a61ee959d9.png)

Reply:
I was just asking because naming can be tricky here due to the `$` and `_` magic prefix handling.

Reply:
Ah yeah, I did check for that.